src/spicy/spicyz/driver.cc

```diff
@@ -31,28 +32,45 @@ using Driver = ::zeek::spicy::Driver;
  * Visitor to type information from a HILTI AST. This extracts user-visible
  * types only, we skip any internal ones.
  */
-struct VisitorTypes : public hilti::visitor::PreOrder<void, VisitorTypes> {
-    explicit VisitorTypes(Driver* driver, hilti::ID module, hilti::rt::filesystem::path path, bool is_resolved)
-        : driver(driver), module(std::move(module)), path(std::move(path)), is_resolved(is_resolved) {}
+struct VisitorTypes : public spicy::visitor::PreOrder {
+    explicit VisitorTypes(Driver* driver, GlueCompiler* glue, bool is_resolved)
+        : driver(driver), glue(glue), is_resolved(is_resolved) {}
+
+    void operator()(hilti::declaration::Module* n) final {
+        if ( n->uid().in_memory ) {
+            // Ignore modules built by us in memory.
+            module = {};
+            return;
+        }
+
+        module = n->scopeID();
+        path = n->uid().path;
 
-    void operator()(const hilti::declaration::Type& t) {
-        assert(! t.type().typeID() || *t.type().typeID() == hilti::ID(module, t.id())); // ensure consistent IDs
+        if ( is_resolved )
+            glue->addSpicyModule(module, path);
+    }
 
-        if ( module == hilti::ID("hilti") || module == hilti::ID("spicy_rt") || module == hilti::ID("zeek_rt") )
+    void operator()(hilti::declaration::Type* n) final {
+        if ( module.empty() || module == hilti::ID("hilti") || module == hilti::ID("spicy_rt") ||
+             module == hilti::ID("zeek_rt") )
             return;
 
+        assert(! n->type()->type()->typeID() ||
+               n->type()->type()->typeID() == hilti::ID(module, n->id())); // ensure consistent IDs
```

Comment:
Yikes, all the `type`s are quite a mouthful 😬

Reply:
It's not as bad as it looks. :-)